### PR TITLE
Check for magic numbers 

### PR DESCRIPTION
**Pull request by ksuderman**
## Why

#11420  
#11425 

## What

Use a prefix search tree (trie) to look for magic numbers at the start of a file before resorting to other methods to determine if a file contains binary content.  The algorithm is now:

1. Use the trie to look for a magic number at the start of the file.
2. Look for binary characters (currently just NULL) in the first 1024 bytes of the file.
3. Look for binary characters in the middle 1024 bytes of the file.

The `os.stat` operation has also been moved so it is only performed if the code needs to seek to the middle of the file. 

As there is no central database of file formats that include magic numbers so the data in `magic-numbers.tsv` was scraped [from the web](https://www.garykessler.net/library/file_sigs.html).

## Todo

- [x] Naming things and package locations.  Is this the correct place this code?
- [x] Test coverage.  ~~I can't find any current tests for the code in `checkers.py`~~
- [x] Remove non-binary file types from `magic-number.tsv` and `magic-numbers.txt` as the current code also identifies XML, OWL, and some HTML files if they start with known strings.
- [x] ~~Clean up and combine the data in `magic-numbers.tsv` and `magic-numbers.txt` so they are easier to maintain.~~
   Will leave as is for now.
- [x] Identify any other binary file types that use magic number but are not included in the list.

## How to test the changes?

(Select all options that apply)

- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License

- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                self.add(number)
+        with open(os.path.join(this_dir, "magic-numbers.txt")) as fp:
+            for line in fp.readlines():
+                parts = line.split('\t')
+                bytes = ' '.join([ hex(ord(c)).replace('0x', '') for c in parts[0]])
+                number = MagicNumber(bytes, parts[1], parts[2])
+                self.add(number)
+
+
+    def add(self, number: MagicNumber):
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]: Should this be a "private" method?
```suggestion
    def _add(self, number: MagicNumber):
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+BLOCK_SIZE=1024
+
+
+class MagicNumber:
+    '''
+    Information about magic numbers loaded from the tsv and text data files.
+    '''
+    def __init__(self, number, extension, description):
+        self.number = number
+        self.bytes = [ int(hex, 16) for hex in number.split() if hex != '??']
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
        self.bytes = [int(hex, 16) for hex in number.split() if hex != '??']
```

File: lib/galaxy/util/checkers.py [hunk 3]
```diff
@@ -8,6 +8,8 @@
     BytesIO,
     StringIO
 )
+#import magic_numbers
+from galaxy.util import magic_numbers
 
 from galaxy import util
 from galaxy.util.image_util import image_type
```
Comment (nsoranzo) [inline lib/galaxy/util/checkers.py]:
```suggestion

from galaxy import util
from galaxy.util import magic_numbers
from galaxy.util.image_util import image_type
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+import os
+import io
+
+# The number of bytes to be read from a file/stream when searching for a
+# magic number.  No magic number is even close to this long, but given most
+# block devices read 4K chunks this is likely a moot point.
+BLOCK_SIZE=1024
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
BLOCK_SIZE = 1024
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        current.formats.append(number)
+
+
+    def sniff(self, path, isFile=True):
+        if isFile:
+            with open(path, 'rb') as f:
+                buff = f.read(BLOCK_SIZE)
+        else:
+            f = io.BytesIO(path)
+            buff = f.read(BLOCK_SIZE)
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
            buff = path
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+class MagicNumberSniffer:
+    '''
+    The prefix search tree (trie) used to identify magic numbers at
+    the start of a file.
+    '''
+    def __init__(self):
+        self.root = TrieNode()
+        this_dir = os.path.dirname(__file__)
+        with open(os.path.join(this_dir, "magic-numbers.tsv")) as fp:
+            for line in fp.readlines():
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
            for line in fp:
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                parts = line.split('\t')
+                number = MagicNumber(*parts)
+                self.add(number)
+        with open(os.path.join(this_dir, "magic-numbers.txt")) as fp:
+            for line in fp.readlines():
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
            for line in fp:
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                parts = line.split('\t')
+                bytes = ' '.join([ hex(ord(c)).replace('0x', '') for c in parts[0]])
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
                bytes = ' '.join(hex(ord(c)).replace('0x', '') for c in parts[0])
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+        # Walk the search tree using the bytes in `buff` as the indices.
+        current = self.root
+        while current is not None:
+            b = buff[p]
+            p += 1
+            if b == '':
+                # EOF
+                break
+            # print(f"ch: {hex(b)}")
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
        for b in buff:
```

File: lib/galaxy/util/checkers.py [hunk 10]
```diff
@@ -50,21 +53,28 @@ def check_binary(name, file_path=True):
...
         if file_path and not is_gzip(name) and not is_zip(name) and not is_bz2(name):
             # file_path=False doesn't seem to be used in the codebase
+            size = os.stat(name).st_size
+            read_start = int(size / 2)
+            temp.seek(read_start)
+            return util.is_binary(temp.read(read_length))
+        if not file_path:
+            read_start = len(name) / 2
             temp.seek(read_start)
             return util.is_binary(temp.read(read_length))
```
Comment (nsoranzo) [inline lib/galaxy/util/checkers.py]: Do we still have a use case for this now that FCS files are handled by magic numbers?

Comment (ksuderman) [reply lib/galaxy/util/checkers.py]: That is a good question.  The problem was exposed by FCS files, but are there other binary formats, without magic numbers, that might exhibit the same problem?  I will run some tests and see which of the binary test files are identified by each part of the code.

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    the start of a file.
+    '''
+    def __init__(self):
+        self.root = TrieNode()
+        this_dir = os.path.dirname(__file__)
+        with open(os.path.join(this_dir, "magic-numbers.tsv")) as fp:
+            for line in fp:
+                parts = line.split('\t')
+                number = MagicNumber(*parts)
+                self.add(number)
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
                self._add(number)
```

File: lib/galaxy/util/magic_numbers/__init__.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        with open(os.path.join(this_dir, "magic-numbers.txt")) as fp:
+            for line in fp:
+                parts = line.split('\t')
+                bytes = ' '.join(hex(ord(c)).replace('0x', '') for c in parts[0])
+                number = MagicNumber(bytes, parts[1], parts[2])
+                self.add(number)
```
Comment (nsoranzo) [inline lib/galaxy/util/magic_numbers/__init__.py]:
```suggestion
                self._add(number)
```

File: lib/galaxy/util/checkers.py (hunk 10 above)
Comment (ksuderman) [reply lib/galaxy/util/checkers.py]: There are no files in the test data (`lib/galaxy/datatypes/test`) that are matched by seeking to the middle of the file; all files binary files are identified (if they are) by either sniffing the magic number or looking for NULL chars in the first 1K characters.  So, it is probably safe to remove this.

Comment (nsoranzo) [reply lib/galaxy/util/checkers.py]: Given the performance issue with the seek, I'd suggest to go ahead with removing it. If we need to restore it at a certain point in the future, it'll still be in the git history.

**Comment by ksuderman** [comment]
This PR should be ready.  The `check_binary` method now accepts the same set of files as the current version of the function.

I rebased my branch and now I see this PR has 4,308 commits to merge... so I may need to redo

**Comment by nsoranzo** [comment]
> This PR should be ready. The `check_binary` method now accepts the same set of files as the current version of the function.
> 
> I rebased my branch and now I see this PR has 4,308 commits to merge... so I may need to redo

I can help with the rebase if you'd like, just let me know when it's a good time.

File: test/unit/util/test_checkers.py [hunk 13]
```diff
@@ -17,3 +22,27 @@ def test_check_html():
         tmpb.write(b'\x1f\x8b')
         tmpb.flush()
         assert not check_html(tmpb.name)
+    with tempfile.NamedTemporaryFile(mode='wb') as tmp:
+        tmp.write(b'\x1f\x8b')
+        tmp.flush()
+        assert not check_html(tmp.name)
```
Comment (nsoranzo) [inline test/unit/util/test_checkers.py]:
```suggestion
    with tempfile.NamedTemporaryFile(mode='wb') as tmpb:
        tmpb.write(b'\x1f\x8b')
        tmpb.flush()
        assert not check_html(tmpb.name)
```

**Comment by ksuderman** [comment]
> Do we still need this ? #13654 also went in a similar direction with the use of python-magic

If the improved `check_binary` is now sniffing for magic numbers then this can likely be closed.  I will leave that call to you.

**Pull request closed without merging**